packages/next/src/server/lib/router-server.ts

```diff
@@ -583,13 +583,19 @@ export async function initialize(opts: {
       })
 
       if (opts.dev && developmentBundler && req.url) {
-        const isHMRRequest = req.url.includes('/_next/webpack-hmr')
+        const { basePath } = config
+
+        const isHMRRequestForHostApp =
+          !basePath && req.url.startsWith('/_next/webpack-hmr')
+
+        const isHMRRequestForGuestApp = pathHasPrefix(req.url, basePath)
+
+        const isHMRRequestForCurrentBasepath =
+          isHMRRequestForHostApp || isHMRRequestForGuestApp
```

Comment:
One of these values would always return true. I have tried this locally and the infinite refresh still happens because of this.

You don't need to check ifs the guestApp, the same check your making for the guestApp just make for the hostApp:

Tried the below locally and it works:
`const { basePath } = config

                const isHMRRequestForHostApp =
                    pathHasPrefix(req.url, basePath) && req.url.startsWith('/_next/webpack-hmr')

                if (isHMRRequestForHostApp) {
                    return developmentBundler.hotReloader.onHMR(req, socket, head)
                }`

Reply:
Hi @jamaybyrone, thanks for your review, I tried to reproduce the thing that you mention in the example app locally and it seems to be working as it expected (without any change). 
Could you provide more details so I can help you?

CMD's that I used to test multi-zone example (provided by Next.js team)
> [!WARNING]  
> If you run the app without `next-with-deps` part the apps will use the current canary version that have the bug.

`pnpm run dev` (For next.js dev mode)
`pnpm next-with-deps ./examples/with-zones/home` (For Home App)
`pnpm next-with-deps ./examples/with-zones/blog` (For Blog App)

> [!IMPORTANT]  
> This PR also updates e2e test cases so it should prevent this behaviour to appear in the future again!

Let me know if I can help you in any way, I will be happy to help!

Video:

https://github.com/vercel/next.js/assets/20648944/224322d1-0aa7-44ce-9d26-8af9d7cba7dd

Reply:
Hmm this is odd....
For me: 
Zone B Calls Zone A

Zone A nextserver is hit:
```
       const isHMRRequestForHostApp =
          !basePath && req.url.startsWith('/_next/webpack-hmr') // returns false

        const isHMRRequestForGuestApp = pathHasPrefix(req.url, basePath) // returns true
```

which then means:
```
const isHMRRequestForCurrentBasepath =
          isHMRRequestForHostApp || isHMRRequestForGuestApp // becomes true in zone A because isHMRRequestForGuestApp is true
```

which then triggers:
```
    if (isHMRRequestForCurrentBasepath) {
          return developmentBundler.hotReloader.onHMR(req, socket, head) // which this to all be hit again in zone A
        }
```

I'm not quite sure how this is working for you isHMRRequestForCurrentBasepath would always be true, there would not be an instant where this is false as its either for the host app or (||) for the guest app.

Reply:
Hi @jamaybyrone, thanks for your answer, could you provide a reproduction environment in [codesandbox](https://codesandbox.io/) or share your next.config.js setup so I can reproduce it locally? (maybe the basePath configuration it's different than the one that was reported with the issue initially #54454)

Reply:
Ahhhhhhh I had:
```
async rewrites() {
    return [
      {
        source: '/:path*',
        destination: `/:path*`,
      }
    ]
```
In the second zone! Sorry for that.

I am still very confused as to when:
isHMRRequestForCurrentBasepath would ever return false, do you have any scenarios for it?

Reply:
isHMRRequestForCurrentBasepath will be true in the host app that contains the next.js rewrites and a default base path (the scenario for it it's the one that I shared in the video) 👍